property.go

```diff
@@ -65,13 +69,41 @@ func ExtractProperty(err error, key Property) (interface{}, bool) {
 }
 
 var (
-	propertyContext = RegisterProperty("ctx")
-	propertyPayload = RegisterProperty("payload")
+	propertyContext    = RegisterProperty("ctx")
+	propertyPayload    = RegisterProperty("payload")
+	propertyUnderlying = RegisterProperty("underlying")
 )
 
 func newProperty(label string) Property {
-	return Property{
-		id:    nextInternalID(),
-		label: label,
+	p := Property{
+		&property{
+			id:    nextInternalID(),
+			label: label,
+		},
+	}
+	return p
+}
+
+// propertyMap represents map of properties.
+// Compared to builtin type, it uses less allocations and reallocations on copy.
+// It is simple binary search tree that relies on property id randomness for balancing.
+type propertyMap struct {
+	p     Property
+	value interface{}
+	next  *propertyMap
+}
+
+// withProperty creates new map with property added or overwritten
+func (pm *propertyMap) with(p Property, value interface{}) *propertyMap {
+	return &propertyMap{p: p, value: value, next: pm}
+}
+
+func (pm *propertyMap) get(p Property) (value interface{}, ok bool) {
+	for pm != nil {
+		if pm.p == p {
```

Comment:
Why do we rely on pointers equality here? Is it in any way better than to compare by id? If not, I would suggest we use id equality here, as it is less confusing and brittle.

Reply:
Equality by id is not better in any way, because pointer equality is just enough in this case.
Equality by pointer is a bit better for the case of one property because no need to fetch property id.

Reply:
OK, let's leave this as pointer equality check; please leave a comment to a private structure to clarify this point, and you can also remove the id from property altogether, as it is actually never used.